### ISSUE TITLE
System freeze after trying to load techcrunch.com

### ISSUE THREAD

**Issue by sviter** [servo/servo#17932]
trying to load https://techcrunch.com on Windows 10 lead to massive memory usage and then system become unresponsive.

**Comment by sviter** [comment]
hi! I don't know much about working with servo, but I was able to build it. After it I open servo.exe in visual studio. After opening https://techcrunch.com/ I've got an exception `thing+1->group_.value was nullptr.` in Marking.cpp in line 783. Is it something useful?

**Comment by jdm** [comment]
Sounds like there's something going wrong in our JS engine integration.

**Comment by jdm** [comment]
```
    frame #0: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] JS::shadow::Zone::runtimeFromMainThread() const at HeapAPI.h:146 [opt]
    frame #1: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] JS::Zone::isGCMarking(this=&0x0) at Zone.h:234 [opt]
    frame #2: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] bool MustSkipMarking<JSObject*>(obj=&0x14a6de700) at Marking.cpp:753 [opt]
  * frame #3: 0x00000001038c3994 servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] void DoMarking<JSObject>(thing=&0x14a6de700) at Marking.cpp:783 [opt]
    frame #4: 0x00000001038c3994 servo`void DispatchToTracer<JSObject*>(trc=&0x118988a28, thingp=<unavailable>, name=&0x1074bee29) at Marking.cpp:651 [opt]
    frame #5: 0x000000010330dce5 servo`::CallObjectTracer(trc=&0x118988a28, objp=&0x125d44d30, name=&0x1074bee29) at jsglue.cpp:828
    frame #6: 0x00000001014a5375 servo`script::dom::bindings::trace::trace_object(tracer=&0x118988a28, description="refcounted", obj=&0x125d44d30) at trace.rs:151
    frame #7: 0x00000001014a511e servo`script::dom::bindings::trace::trace_reflector(tracer=&0x118988a28, description="refcounted", reflector=&0x125d44d30) at trace.rs:144
    frame #8: 0x00000001014a299b servo`script::dom::bindings::refcounted::trace_refcounted_objects::{{closure}}(r=&0x118966468) at refcounted.rs:297
    frame #9: 0x00000001011d4eed servo`std::thread::local::{{impl}}::try_with<alloc::rc::Rc<core::cell::RefCell<core::option::Option<script::dom::bindings::refcounted::LiveDOMReferences>>>,closure,()>(self=&0x1087bc348, f=closure(&0x7000327f22c0)) at local.rs:362
    frame #10: 0x0000000100ef3f3d servo`std::thread::local::{{impl}}::with<alloc::rc::Rc<core::cell::RefCell<core::option::Option<script::dom::bindings::refcounted::LiveDOMReferences>>>,closure,()>(self=&0x1087bc348, f=closure(&0x7000327f22c0)) at local.rs:276
    frame #11: 0x00000001014a2711 servo`script::dom::bindings::refcounted::trace_refcounted_objects(tracer=&0x118988a28, _data=&0x0) at refcounted.rs:289
    frame #12: 0x00000001038ce732 servo`js::gc::GCRuntime::markRuntime(this=&0x118986bc8, trc=&0x118988a28, traceOrMark=<unavailable>, lock=<unavailable>) at RootMarking.cpp:343 [opt]
    frame #13: 0x00000001036c2cf1 servo`js::gc::GCRuntime::beginMarkPhase(this=&0x118986bc8, reason=<unavailable>, lock=<unavailable>) at jsgc.cpp:3907 [opt]
    frame #14: 0x00000001036c8a45 servo`js::gc::GCRuntime::incrementalCollectSlice(this=&0x118986bc8, budget=<unavailable>, reason=DESTROY_RUNTIME, lock=<unavailable>) at jsgc.cpp:5889 [opt]
    frame #15: 0x00000001036c939f servo`js::gc::GCRuntime::gcCycle(this=&0x118986bc8, nonincrementalByAPI=<unavailable>, budget=0x00007000327f25c0, reason=DESTROY_RUNTIME) at jsgc.cpp:6174 [opt]
    frame #16: 0x00000001036c98ef servo`js::gc::GCRuntime::collect(this=&0x118986bc8, nonincrementalByAPI=true, budget=<unavailable>, reason=DESTROY_RUNTIME) at jsgc.cpp:6287 [opt]
    frame #17: 0x00000001036c9c01 servo`js::gc::GCRuntime::gc(this=<unavailable>, gckind=<unavailable>, reason=<unavailable>) at jsgc.cpp:6354 [opt]
    frame #18: 0x00000001037b61b8 servo`JSRuntime::destroyRuntime(this=&0x1189867c8) at Runtime.cpp:430 [opt]
    frame #19: 0x000000010368be44 servo`js::DestroyContext(JSContext*) [inlined] JSContext::~JSContext(this=&0x118986600) at jscntxt.cpp:892 [opt]
    frame #20: 0x000000010368be3c servo`js::DestroyContext(JSContext*) [inlined] JSContext::~JSContext() at jscntxt.cpp:891 [opt]
    frame #21: 0x000000010368be3c servo`js::DestroyContext(JSContext*) [inlined] void js_delete_poison<JSContext>(p=<unavailable>) at Utility.h:392 [opt]
    frame #22: 0x000000010368be2f servo`js::DestroyContext(cx=&0x118986600) at jscntxt.cpp:133 [opt]
    frame #23: 0x0000000103305f5d servo`js::rust::{{impl}}::drop(self=&0x12025a010) at rust.rs:240
    frame #24: 0x0000000101f3c205 servo`core::ptr::drop_in_place<js::rust::Runtime>((null)=&0x12025a010) at ptr.rs:61
    frame #25: 0x0000000101f4adc9 servo`core::ptr::drop_in_place<alloc::rc::RcBox<js::rust::Runtime>>((null)=&0x12025a000) at ptr.rs:61
    frame #26: 0x0000000101fc5ee9 servo`alloc::rc::{{impl}}::drop<js::rust::Runtime>(self=&0x7000327f2c08) at rc.rs:714
    frame #27: 0x0000000101340cd5 servo`core::ptr::drop_in_place<alloc::rc::Rc<js::rust::Runtime>>((null)=&0x7000327f2c08) at ptr.rs:61
    frame #28: 0x00000001013365c7 servo`core::ptr::drop_in_place<script::script_thread::ScriptThread>((null)=&0x7000327f2a70) at ptr.rs:61
    frame #29: 0x000000010152bfe6 servo`script::script_thread::{{impl}}::create::{{closure}} at script_thread.rs:605
```

```
frame #0: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] JS::shadow::Zone::runtimeFromMainThread() const at HeapAPI.h:146 [opt]
   143
   144 	    JSRuntime* runtimeFromMainThread() const {
   145 	        MOZ_ASSERT(js::CurrentThreadCanAccessRuntime(runtime_));
-> 146 	        return runtime_;
   147 	    }
   148
   149 	    // Note: Unrestricted access to the zone's runtime from an arbitrary
(lldb) up
frame #1: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] JS::Zone::isGCMarking(this=&0x0) at Zone.h:234 [opt]
   231 	    }
   232
   233 	    bool isGCMarking() {
-> 234 	        if (runtimeFromMainThread()->isHeapCollecting())
   235 	            return gcState_ == Mark || gcState_ == MarkGray;
   236 	        else
   237 	            return needsIncrementalBarrier();
(lldb) up
frame #2: 0x00000001038c39bc servo`void DispatchToTracer<JSObject*>(JSTracer*, JSObject**, char const*) [inlined] bool MustSkipMarking<JSObject*>(obj=&0x14a6de700) at Marking.cpp:753 [opt]
   750 	    // Don't mark things outside a zone if we are in a per-zone GC. It is
   751 	    // faster to check our own arena, which we can do since we know that
   752 	    // the object is tenured.
-> 753 	    return !TenuredCell::fromPointer(obj)->zone()->isGCMarking();
   754 	}
   755
   756 	template <>
```

So zone() is returning a null pointer here.

**Comment by jdm** [comment]
https://github.com/jdm/servo/commit/359bd75b06cdaae80a4078c9f2892aa6867c89f8 fixes this issue for me, but I'm having trouble writing an automated test that demonstrates the issue.

**Comment by sviter** [comment]
No crash now. But after loading this page I can see constantly high using of CPU

**Comment by jdm** [comment]
The crash should have been fixed by #18952. Let's file a separate issue for high CPU usage.
